### PR TITLE
[ML] Add datafeed_config field to anomaly detection job configs

### PR DESCRIPTION
**Pull request by benwtrent**
This is a quality of life improvement for typical users. Almost all anomaly detection jobs receive their input data through a datafeed. 

The datafeed config can now be supplied and is available in the `datafeed_config` field in the job config for creation and getting jobs.

### CODE REVIEW

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutJobAction.java
Comment (przemekwitek) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutJobAction.java:67]:
```suggestion
                if (jobCreated.getResponse().getDatafeedConfig().isEmpty()) {
```

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutJobAction.java
Comment (przemekwitek) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutJobAction.java:79]:
```suggestion
                        datafeedCreated -> listener.onResponse(jobCreated),
```
?
This will instantly give the reader a clue what is created.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutJobAction.java
Comment (przemekwitek) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutJobAction.java:88]: If we fail here, will we have some maintenance activity cleaning those jobs up?
Or will it be a responsibility of the user?

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/DatafeedManager.java
Comment (przemekwitek) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/DatafeedManager.java:205]: Are all the datafeeds stored in cluster state? Or is this code only to serve the legacy datafeeds (created before ML configs were moved from cluster state to index)?

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedConfigBuilderTests.java
Comment (przemekwitek) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedConfigBuilderTests.java:17]: Should this method be moved into this class (`DatafeedConfigBuilderTests`)?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedConfig.java
Comment (przemekwitek) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedConfig.java:775]:
```suggestion
```

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/JobManager.java
Comment (przemekwitek) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/JobManager.java:230]:
```suggestion
            jobBuilders -> jobsListener.onResponse(
```
?

File: docs/reference/ml/anomaly-detection/apis/get-job.asciidoc
Comment (droberts195) [inline docs/reference/ml/anomaly-detection/apis/get-job.asciidoc:84]: I am thinking it might make UI integration and consistency of what users see during upgrades better if this is called `datafeed_config`.

At the moment the UI already pretends that the datafeed config is nested inside the job config in the JSON panel of the jobs list, but nests it in an object named `datafeed_config`, not `datafeed`.

So for what the user sees to remain consistent it would be better if the nested object added to the job config by the backend code was called `datafeed_config`.  Otherwise in the period when the backend might be returning `datafeed` and the UI is still getting the datafeed separately the JSON panel in the jobs list in the UI will show both `datafeed` _and_ `datafeed_config`.

The one potential problem here would be if the UI currently throws an error if a job config JSON already contains `datafeed_config` instead of simply overwriting it with a separately obtained `datafeed_config`.  If the UI throws an error in this situation then we'll have to stick with `datafeed` and tolerate the duplication in versions of the UI that are already released.  If the UI throws an error when `datafeed_config` is already present we cannot fix this now, as the upgrade rules say that Elasticsearch must be upgraded before Kibana, and so already released UI code from 7.0 to 7.13 must work with 7.14 - 7.last Elasticsearch.

/cc @jgowdyelastic @peteharverson

File: docs/reference/ml/anomaly-detection/apis/get-job.asciidoc
Comment (benwtrent) [reply docs/reference/ml/anomaly-detection/apis/get-job.asciidoc:90]: @przemekwitek 
If you are just worrying about the `datafeed_id` field, I am not to worried about that. Its the "unique id of the datafeed configuration". If it was a more nuanced and complicated field, moving it to the shared file may be necessary.

I will leave the final call to @szabosteve or @lcawl

File: x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/ml/jobs_crud.yml
Comment (benwtrent) [reply x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/ml/jobs_crud.yml:1026]: it doesn't, I am guessing the test is failing :)

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/DatafeedManager.java
Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/DatafeedManager.java:205]: This is legacy only. Datafeeds are supposed to be fully migrated into the configuration index.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutJobAction.java
Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutJobAction.java:88]: > maintenance activity cleaning those jobs up?

There is no way of knowing if the job was initially created with a datafeed or not. I don't think it wise to have a maintenance action to delete any job that doesn't have a datafeed.

The user will have to intervene here and it SHOULD be a rare occurrence.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutJobAction.java
Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutJobAction.java:67]: I am using `.isPresent() == false` for backportability.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/job/config/JobTests.java
Comment (benwtrent) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/job/config/JobTests.java:91]: Ah, I need to delete this method.

File: docs/reference/ml/anomaly-detection/apis/get-job.asciidoc
Comment (benwtrent) [reply docs/reference/ml/anomaly-detection/apis/get-job.asciidoc:84]: I am fine changing it to `datafeed_config`. 

> The one potential problem here would be if the UI currently throws an error if a job config JSON already contains `datafeed_config`

That would be a problem indeed. Eagerly awaiting feed back @jgowdyelastic @peteharverson :)

**Comment by droberts195** [comment]
I will have a closer look at this later.  I don't think it should go into 7.14 though.  I would rather hold it for 7.15 to give time to iron out any unforeseen side effects.

File: docs/reference/ml/anomaly-detection/apis/get-job.asciidoc
Comment (szabosteve) [reply docs/reference/ml/anomaly-detection/apis/get-job.asciidoc:90]: It is very unlikely that the description of `datafeed-id` will change dramatically, but as we already have this property added to the ml-shared file, I make a suggestion below to use it here.

File: docs/reference/ml/anomaly-detection/apis/get-job.asciidoc
Comment (szabosteve) [inline docs/reference/ml/anomaly-detection/apis/get-job.asciidoc:92]:
```suggestion
include::{es-repo-dir}/ml/ml-shared.asciidoc[tag=datafeed-id]
```

File: docs/reference/ml/anomaly-detection/apis/get-job.asciidoc
Comment (szabosteve) [inline docs/reference/ml/anomaly-detection/apis/get-job.asciidoc:85]:
```suggestion
(object) The {dfeed} configured for the current {anomaly-job}.
```

**Review by droberts195 — APPROVED** [review]
I checked with @jgowdyelastic and he thought that the jobs list would work OK with `datafeed_config` already existing.  It might overwrite the field or might just not add it if already present, but either is fine.  However, it would be best to test it to be sure, and given the way the UI master branch is now this can be tested on the master branch.

However, one other problem he thought of was that if `exclude_generated` doesn't exclude the `datafeed_config` then it will break the clone functionality in the UI in the short term.  The way this PR is currently implemented, getting jobs will never include the datafeed config, as that part is not persisted to the index, so this PR won't break clone.  However, eventually we'll want to store the datafeed config together with the job config so that we can more efficiently retrieve both together.

Given the uncertainty around the Kibana jobs list I think next steps should be:

1. Merge this PR to a feature branch
2. Create another PR on top of it that persists the datafeed config to the `.ml-config` index
3. Confirm that in that second PR, the get jobs endpoint shows datafeed configs for those jobs that contain datafeeds (I know this PR will have lots of inconsistencies like update datafeed only changing the standalone datafeed, but ignore that for now)
4. Manually test master branch Kibana locally using `yarn start` against the second PR on the feature branch
5. Confirm that there are no errors or crashes as a result of get jobs returning jobs that already contain a datafeed config to the UI

If there _are_ errors or crashes in step 5 then we need to rethink either the field naming or sequencing of changes, hence why this PR should only go on a feature branch initially.

If the the master branch UI works fine against that second PR then the first PR on the feature branch (i.e. this one) can safely go to master and 7.15.

So, LGTM to merge this PR to a feature branch.

**Comment by benwtrent** [comment]
> The way this PR is currently implemented, getting jobs will never include the datafeed config, as that part is not persisted to the index, 

The `datafeed_config` field is indeed included on `GET job`, and is also included in `exclude_generated`. I will test this particular PR in the way suggested.

> However, eventually we'll want to store the datafeed config together with the job config so that we can more efficiently retrieve both together.

This is indeed a possibility, but not a foregone conclusion (at least not to me). But, as it stands now, since this PR includes the datafeed_config on `get` its a facsimile of storing the config along with the job in the index :).

Let me test against kibana and report back.

**Comment by benwtrent** [comment]
OK, Kibana wizards work except for `clone`

image

I am guessing that since clone calls `get` and then `put`, there is an issue with the Typescript definition. 

@jgowdyelastic ^ is there a way to make this BWC with Kibana so that older kibana versions don't barf with clone? Or do we not support `clone` when the version of ES is newer?

**Comment by jgowdyelastic** [comment]
>@jgowdyelastic ^ is there a way to make this BWC with Kibana so that older kibana versions don't barf with clone? Or do we not support clone when the version of ES is newer?

We're always going to have a problem if there is a newer es with new properties in a job config which are not expected by our kibana schemas. We've not hit this so far, as far as I know.
Us having schema checks on the kibana endpoints which are just basic wrappers for es endpoints (like create job) aren't that helpful. and are possibly more of a hinderance.
If the payload is wrong, es will quickly tell us.

the `include_datafeed ` flag might be a good solution for this.
We could also add in a line to remove `datafeed_config` when cloning.

**Comment by droberts195** [comment]
> is there a way to make this BWC with Kibana so that older kibana versions don't barf with clone?

I think the way to do that would be _not_ to include `datafeed_config` when `exclude_generated` was `true`.

However, that is at odds with the long term direction of this work.

There are basically two options:

1. Make the 7.last UI tolerate `datafeed_config` either present or missing during clone (since 7.last is the only 7.x release where Kibana needs to know how to talk to an 8.x Elasticsearch as well as 7.last), change 8.0 Elasticsearch to return `datafeed_config` when `exclude_generated` is `true` and leave 7.x Elasticsearch not returning `datafeed_config` when `exclude_generated` is `true`.  (Effectively saying `datafeed_config` is a "generated" field in 7.x.)
2. Accept that clone doesn't work in the partly upgraded state where Kibana is on pre-7.15 and Elasticsearch on post-7.15.  It might not be a big issue because people should be focusing on completing their upgrade before using new functionality.

During the course of 7.15 we need to think through all the other breaking steps we need to make, and whether we can do them all on the 7.last/8.0 boundary or whether we need two major releases to do them.

**Comment by benwtrent** [comment]
@droberts195 I think the `exclude_generated` behavior change between 7.x and 8.x makes sense. This way kibana has time in 7.series to do the correct thing (allowing it to be around) and in 8.x we simply say its always around.

**Comment by benwtrent** [comment]
@jgowdyelastic @droberts195 ok, now the `datafeed_config` is not included on `get jobs` when `exclude_generated: true`. Once this is merged in, we can adjust kibana to allow a `datafeed_config` to be optionally allowed on `GET` and update version 8 ES to then include it on `exclude_generated: true`.

**Comment by droberts195** [comment]
> ok, now the `datafeed_config` is not included on `get jobs` when `exclude_generated: true`.

OK, I'm happy for this to be merged to master and 7.15 then.  But please don't put it in 7.14, just in case there are further unforeseen side effects.

**Review by szabosteve — APPROVED** [review]
Docs LGTM, thanks!

**Review by przemekwitek — APPROVED** [review]
LGTM

**Pull request merged**
